### PR TITLE
Third cumulant

### PR DESCRIPTION
**Pull request by DnMGalan**
**Description**

Adding the first three cumulants using the expressions of C. Flindt's PhD Thesis (page 80): https://orbit.dtu.dk/en/publications/electrons-in-nanostructures-coherent-manipulation-and-counting-st

I have modified the code so it is consistent with the previous work made in QuTip. And I have added it as an independent function so it doesn't modify the original code.

@nwlambert these are the changes mentioned in my previous email.

- [ ] The changelog has yet to be added

**Related issues or PRs**
feature #442

### CODE REVIEW

**Comment by nwlambert** [comment]
Thanks @DnMGalan !  can you make a test for it? 

Also could you share an example for the issue with the small finite frequency?  historically it was there to fix an some edge cases that fail at w=0, but sounds like you found the opposite?

**Comment by DnMGalan** [comment]
Hi @nwlambert,

Sure! This is the code I used to test the changes:

```
import numpy as np
from qutip import *

Gamma_r = 0.5
Gamma_l = 0.1

n_ket = (1)/(Gamma_l + Gamma_r)*Qobj([[Gamma_r],[Gamma_l]], )
n_bra = Qobj([[1, 1]])

Id = Qobj([[1, 0], [0, 1]])

L = Qobj([[-Gamma_l, Gamma_r], [Gamma_l, -Gamma_r]],)

I_ops = Qobj([[0, Gamma_r], [0, 0]])

L_s = sprepost(L, Id)

I_s = sprepost(I_ops, Id)

current, noise, skw = third_cumulant(L_s, rhoss=n_ket*n_bra, J_ops=[I_s], I_ops=[I_s], sparse=True, method="pinv")

print(current, noise, skw)
```
It is based in a simple analytical example I found in a paper, which contains the expressions to check the results. This is the paper (page 7, the section "Toy model"): [here is the paper](https://orbit.dtu.dk/en/publications/current-and-current-fluctuations-in-quantum-shuttles).

As you can see it is really simple, although I had to make some changes to be able to applied in QuTip (I hope this doesn't lead to major fails).

The effect of the frequency for the case when it is not defined in the function can be observed using the same example, just don't modify that part and you will see it, then you can try the solution I propose ( I don't know if I explained myself correctly).

**Comment by nwlambert** [comment]
thanks Dani, I think if you encode the problem in a more standard qutip structure the dependence on the small frequencies disappears;
```
d = destroy(2).dag()
L_s = liouvillian(0*d.dag()*d, [np.sqrt(Gamma_l) *d, np.sqrt(Gamma_r) *d.dag()])
rho_ss = steadystate(L_s)
I_s = Gamma_r * sprepost(d.dag(),d)
current, noise, skw = third_cumulant(L_s, rhoss=rho_ss, J_ops=[I_s], I_ops=[I_s], sparse=True, method="pinv")
```

So I would prefer to revert the pseudo_inverse change (which is there originally to catch a different issue with the noise exploding).  But nice example!  this would make a great test I think

For the third_cumulant itself, maybe better to combine with the existing countstat_current_noise function, and provide the third cumulant, and support J_ops!=I_ops, as options?

**Comment by DnMGalan** [comment]
Yes! You are right, your example solves the problem with the frequency, I'm sorry seems I addapted wrong the paper to qutip. 

I agree, that is better to let it as it was.

Regarding the third cumulant, that was my idea but first I wanted to show you the code before implementing it in the function countstat_current_noise, I was thinking of adding it as part of the _noise_pseudoinv function, if you are ok with that?

**Comment by nwlambert** [comment]
sounds good to me!  is a direct-like solution also possible?  the multiple R terms make it tricky i guess?

**Comment by DnMGalan** [comment]
Exactly, what do you mean by direct-like solution? I don't if you have something in mind.

**Comment by nwlambert** [comment]
Just that we also have '_noise_direct()' which does a bit of linear algebra to avoid using the pseudo-inverse.  i think it can be a bit more numerically efficient than mucking around with the inverse itself.  Maybe doing the pseudo-inverse tends to involve a big large dense matrix, and this avoids it?

iirc, the logic of noise_direct is, since we write the pseudo-inverse like $R=QL^{-1}Q$, the annoying term we have to evaluate in the noise expression is something like:
$$y = I QL^{-1} Q I \rho.   $$
so we substitute, for the last few terms
$$x = L^{-1} QI\rho $$
we multiply from the left by L, to get
$$L x = Q I \rho$$
and we can find x using a scipy's solve or something equivalent, and then find y with
$$ y = IQ x$$

In the skewness you have this new term,
$$y =I R R I P I \rho$$
which is a bit clunky to manipulate.
maybe we can write something like
$$x = L^{-1}QIPI \rho$$
and again, multiplying by L
$$Lx = QIPI \rho$$
to get x, then the original expression is now

$$y = IQL^{-1}QQx$$
defining a new variable
$$t = L^{-1}QQx$$
we can again, multiply by L
$$Lt = QQx$$,
solve for t, then we get 
$$y=I R R I P I \rho = IQ t$$

? i probably made a mistake, so take with a grain of salt!

**Comment by DnMGalan** [comment]
Ahhhhhh I see, (I completly forgot about the _noise_direct function) yes that sound interesting, I can also try it.

**Comment by DnMGalan** [comment]
I will update this pull request so that it is ready for review, although, I think maybe some changes in the code style will be needed.

One thing I would like to mention is that the _noise_direct function seems to yield a more accurate result than the noise_pseudoinv function, this may be due to the selected method (for the case of noise_pseudoinv), the discussion we had at the beginning about not including the low value of the frenquency or some mistake I made testing? In any case I just wanted to point out that even though both functions work, noise_direct gives a more accurate result to the analytical one.

File: qutip/solver/countstat.py [hunk 1]
```diff
@@ -107,38 +114,77 @@ def _noise_direct(L, wlist, rhoss, J_ops):
...
+            op = I_ops[i] @ R @ I_ops[j] + I_ops[j] @ R @ I_ops[i]
             noise[i, j, k] -= op(rhoss).tr().real
-    return current, noise
+
+    return current, noise, skewnes
 
 
-def countstat_current_noise(L, c_ops, wlist=None, rhoss=None, J_ops=None,
-                            sparse=True, method='direct'):
+def countstat_current_noise(L, c_ops, wlist=None, rhoss=None, J_ops=None,
```
[the remark this answers is not in the document]
Comment (nwlambert) [reply qutip/solver/countstat.py:186]: ah, in that docstring, good to mention that while noise supports cross-correlations between different J_ops,  skewness doesn't, and just returns diagonal (self-correlated) terms.

**Review by nwlambert — COMMENTED** [review]
Thanks again for working on this, sorry it fell through the cracks.

Left a few initial docstring comments.   Maybe good to double check pep8.   Some of the definitions in the direct solver couldn't be polished/compacted a bit, but not sure the best approach.

I quickly tried the analytical example again, so that we can use as a test, but the skewness itself doesn't quite match up with that analytical result, not sure why, will updated if I spot it (I probably made a typo inputting the formula or something)

```
import numpy as np
from qutip import *

Gamma_r = 0.5
Gamma_l = 0.1
d = destroy(2).dag()
L_s = liouvillian(0*d.dag()*d, [np.sqrt(Gamma_l) *d, np.sqrt(Gamma_r) *d.dag()])
rho_ss = steadystate(L_s)
I_s = Gamma_r * sprepost(d.dag(),d)
current, noise, skw = countstat_current_noise(L_s, [], rhoss=rho_ss, J_ops=[I_s], I_ops=[I_s], sparse=True, method="pinv")
print(current, noise, skw)
current, noise, skw = countstat_current_noise(L_s, [], rhoss=rho_ss, J_ops=[I_s], I_ops=[I_s], sparse=True)
print(current, noise, skw)

current_ana = Gamma_l* Gamma_r/(Gamma_l+Gamma_r)
print(current_ana)
noise_ana = current_ana * (Gamma_r**2 + Gamma_l**2)/(Gamma_r+Gamma_l)**2
print(noise_ana)
skewness_ana = current_ana*(Gamma_r**4 - 2*Gamma_r**3*Gamma_l + 6*Gamma_r**2*Gamma_l**2-2*Gamma_r*Gamma_l**3 + Gamma_l**4)/(Gamma_r+Gamma_l)**2
print(skewness_ana)
```

**Comment by nwlambert** [comment]
I had a chat with christian and we stared at it a bit, seems there's a typo in the  skewness in equation 27 of https://pubs.aip.org/aip/pof/article/17/10/100613/985430/Current-and-current-fluctuations-in-quantum, it should be

`skewness_ana = current_ana*(Gamma_r**4 - 2*Gamma_r**3*Gamma_l + 6*Gamma_r**2*Gamma_l**2-2*Gamma_r*Gamma_l**3 + Gamma_l**4)/(Gamma_r+Gamma_l)**4`

power of 4 in the denominator instead of 2.  this calculation is also in his thesis, without the typo, equation 4.30, page 63.

Anyways, key thing is the numerics methods both give same result, and we can use the above as a test! I think just tidying up the doctstrings and adding the test should be enough, and we can merge

**Comment by DnMGalan** [comment]
Hi Neil,

Thank you for the updates. I have some ideas to improve the definitions of the direct solver.

Should I add the test to an existing file, or would you prefer that I create a new one? (Interestingly, I realized that I had already corrected this in the script I used for local testing.

Regarding the changes to the docstring: in the case where I_ops is None, we assume that I−=0, allowing us to set I_ops = J_ops, similarly, when J_ops is None, it  is calculated using the collapse operators. I was considering adding these explanations to the docstrings, along with the definitions based on the currents across a junction that you mentioned. However, I’m not sure if you’re expecting a more general definition.

**Comment by nwlambert** [comment]
For the tests, I think you can just add it to the existing test_countstat file:
https://github.com/qutip/qutip/blob/master/qutip/tests/solver/test_countstat.py

>I was considering adding these explanations to the docstrings, along with the definitions based on the currents across a junction >that you mentioned. However, I’m not sure if you’re expecting a more general definition.

Sounds ok to me!

File: qutip/solver/countstat.py [hunk 2]
```diff
@@ -170,7 +258,14 @@ def countstat_current_noise(L, c_ops, wlist=None, rhoss=None, J_ops=None,
...
     J_ops : array / list (optional)
-        List of current superoperators.
+        List of total current superoperators. Where J_ops is defined as the
+        sum of currents across a junction J_ops ≡ I+ + I-. If not provided,
+        it will be calculated using the collapse operators. 
+
+    I_ops : array / list (optional)
+        List of net current superoperators. Where I_ops is defined as the
+        difference of currents across a junction I_ops ≡ I+ - I-. If not provided,
+        it will be asumed the case I- = 0 and set I_ops = J_ops.
```
Comment (nwlambert) [inline qutip/solver/countstat.py:268]: small typo here "assumed that I- =..."

**Comment by nwlambert** [comment]
Thanks again dani, just a small typo, if you do a  quick commit and run the tests again I will merge (sorry for the delay)

**Comment by DnMGalan** [comment]
No worries. It should be fixed now.

**Comment by nwlambert** [comment]
Thanks Dani, one last quick request.  some tests are failing maybe just because the tolerance on the tests are so high, i guess we can just reduce them a bit.     also, could you do one last check of pep8 with black or something similar to make the codeclimate happy (just some line lengths and indentations I think).   then I will merge

**Comment by DnMGalan** [comment]
Yes, of course!

Is there a reason why the test worked on my machine but failed in the pull request CI? I guess it might be due to differences in the setup (Python version, OS...)

**Comment by DnMGalan** [comment]
I’ve just pushed the changes. I’ve checked everything locally, but it’s still possible that some of the online tests may fail. I’ll review the results once they finish running and make any further adjustments if needed.

**Comment by DnMGalan** [comment]
I made another push to resolve the remaining style issues. This time, I used a different library to detect the errors (since Black didn’t work well). I hope there are no more style errors now. In any case, if there are any issues with the tests, I will fix them.

**Review by nwlambert — APPROVED** [review]
I think the tolerances thing is ok, I tried it locally and with pseudoinverse there was some small difference in the noise around a tolerance 1e-7,  which I guess could be attributed to tolerances in the linear solver or pseudo inverse.  1e-4 seems reasonable for the magnitude of the values you get with the example you use.
